src/qs8-vadd/gen/qs8-vadd-minmax-hvx-u128.c

```diff
@@ -0,0 +1,182 @@
+    HVX_Vector va1_i16_lo = Q6_V_lo_W(Q6_Wh_vunpack_Vb(va1));
+    HVX_Vector vb1_i16_lo = Q6_V_lo_W(Q6_Wh_vunpack_Vb(vb1));
+    HVX_Vector va2_i16_lo = Q6_V_lo_W(Q6_Wh_vunpack_Vb(va2));
+    HVX_Vector vb2_i16_lo = Q6_V_lo_W(Q6_Wh_vunpack_Vb(vb2));
+    HVX_Vector va3_i16_lo = Q6_V_lo_W(Q6_Wh_vunpack_Vb(va3));
+    HVX_Vector vb3_i16_lo = Q6_V_lo_W(Q6_Wh_vunpack_Vb(vb3));
+
+    // vacc = vbias + va * va_multiplier + vb * vb_multiplier
+    HVX_Vector va0_mul_e = Q6_Vw_vmpyio_VwVh(va_multiplier, Q6_Vh_vshuffe_VhVh(va0_i16_lo, va0_i16_lo));
+    HVX_Vector vb0_mul_e = Q6_Vw_vmpyio_VwVh(vb_multiplier, Q6_Vh_vshuffe_VhVh(vb0_i16_lo, vb0_i16_lo));
+    HVX_Vector va0_mul_o = Q6_Vw_vmpyio_VwVh(va_multiplier, va0_i16_lo);
+    HVX_Vector vb0_mul_o = Q6_Vw_vmpyio_VwVh(vb_multiplier, vb0_i16_lo);
+    HVX_VectorPair va0_mul = Q6_W_vshuff_VVR(va0_mul_o, va0_mul_e, -4);
+    HVX_VectorPair vb0_mul = Q6_W_vshuff_VVR(vb0_mul_o, vb0_mul_e, -4);
+    HVX_Vector vacc0 = Q6_Vw_vadd_VwVw(vbias, Q6_V_lo_W(va0_mul));
+    vacc0 = Q6_Vw_vadd_VwVw(vacc0, Q6_V_lo_W(vb0_mul));
+    HVX_Vector va1_mul_e = Q6_Vw_vmpyio_VwVh(va_multiplier, Q6_Vh_vshuffe_VhVh(va1_i16_lo, va1_i16_lo));
+    HVX_Vector vb1_mul_e = Q6_Vw_vmpyio_VwVh(vb_multiplier, Q6_Vh_vshuffe_VhVh(vb1_i16_lo, vb1_i16_lo));
+    HVX_Vector va1_mul_o = Q6_Vw_vmpyio_VwVh(va_multiplier, va1_i16_lo);
+    HVX_Vector vb1_mul_o = Q6_Vw_vmpyio_VwVh(vb_multiplier, vb1_i16_lo);
+    HVX_VectorPair va1_mul = Q6_W_vshuff_VVR(va1_mul_o, va1_mul_e, -4);
+    HVX_VectorPair vb1_mul = Q6_W_vshuff_VVR(vb1_mul_o, vb1_mul_e, -4);
+    HVX_Vector vacc1 = Q6_Vw_vadd_VwVw(vbias, Q6_V_lo_W(va1_mul));
+    vacc1 = Q6_Vw_vadd_VwVw(vacc1, Q6_V_lo_W(vb1_mul));
+    HVX_Vector va2_mul_e = Q6_Vw_vmpyio_VwVh(va_multiplier, Q6_Vh_vshuffe_VhVh(va2_i16_lo, va2_i16_lo));
+    HVX_Vector vb2_mul_e = Q6_Vw_vmpyio_VwVh(vb_multiplier, Q6_Vh_vshuffe_VhVh(vb2_i16_lo, vb2_i16_lo));
+    HVX_Vector va2_mul_o = Q6_Vw_vmpyio_VwVh(va_multiplier, va2_i16_lo);
+    HVX_Vector vb2_mul_o = Q6_Vw_vmpyio_VwVh(vb_multiplier, vb2_i16_lo);
+    HVX_VectorPair va2_mul = Q6_W_vshuff_VVR(va2_mul_o, va2_mul_e, -4);
+    HVX_VectorPair vb2_mul = Q6_W_vshuff_VVR(vb2_mul_o, vb2_mul_e, -4);
+    HVX_Vector vacc2 = Q6_Vw_vadd_VwVw(vbias, Q6_V_lo_W(va2_mul));
+    vacc2 = Q6_Vw_vadd_VwVw(vacc2, Q6_V_lo_W(vb2_mul));
+    HVX_Vector va3_mul_e = Q6_Vw_vmpyio_VwVh(va_multiplier, Q6_Vh_vshuffe_VhVh(va3_i16_lo, va3_i16_lo));
+    HVX_Vector vb3_mul_e = Q6_Vw_vmpyio_VwVh(vb_multiplier, Q6_Vh_vshuffe_VhVh(vb3_i16_lo, vb3_i16_lo));
+    HVX_Vector va3_mul_o = Q6_Vw_vmpyio_VwVh(va_multiplier, va3_i16_lo);
+    HVX_Vector vb3_mul_o = Q6_Vw_vmpyio_VwVh(vb_multiplier, vb3_i16_lo);
+    HVX_VectorPair va3_mul = Q6_W_vshuff_VVR(va3_mul_o, va3_mul_e, -4);
+    HVX_VectorPair vb3_mul = Q6_W_vshuff_VVR(vb3_mul_o, vb3_mul_e, -4);
+    HVX_Vector vacc3 = Q6_Vw_vadd_VwVw(vbias, Q6_V_lo_W(va3_mul));
+    vacc3 = Q6_Vw_vadd_VwVw(vacc3, Q6_V_lo_W(vb3_mul));
+
+    // right shift
```

Comment:
Can you use vasrwhrndsat here? It does a shift right and saturating narrow in one instruction.

Reply:
So, I tried and I found the results was not correct.. I suspect it's because of the way vasrwhsat put the results back into the destination vector: "Each even element in the destination vector register Vd comes from the vector register Vv, and each odd element in Vd comes from the vector register Vu." 

So the first input vector will be placed in odd position, and the second one will be in even position. As I need to keep the original order, this may not work for this case. Do you have any suggsetions?

Reply:
When programming Hexagon, there's a bit of a puzzle to solve when doing widening and narrowing operations like this. Almost always, there is a set of instructions you can use such that this interleaving/deinterleaving fully cancels out, so you don't need to deal with the reordering. In this case, there are two different "unpack" instructions: there is `unpack`, and also `sxt`/`zxt`. Halide would choose between these two instructions depending on which one avoided the extra interleave/deinterleave operation: https://github.com/halide/Halide/blob/6c8a4911c664e645bd4f37d846d1eb2515d12a9b/src/HexagonOptimize.cpp#L1976-L1983

So, my guess is that if you were to replace your `unpack` ops with `sxt` ops, you could use this instruction, and the reorderings would cancel out.

Reply:
Thank you for the information here, very useful! I will review them and try it out again.

Reply:
I ended up using vasrwhsat but shift was not working as I intended. I found vasrwhsat clamps the shift value between 0 and 15 (code below), where our shift value is clearly larger than 16 many cases.

```
// vasrwhsat
// Vd.h=vasr(Vu.w,Vv.w,Rt)[:sat]
for (i = 0; i < VELEM(32); i++) {
  **shamt = Rt & 0xF;**
  Vd.w[i].h[0]=[sat_16](Vv.w[i] >> shamt);
  Vd.w[i].h[1]=[sat_16](Vu.w[i] >> shamt) ;
}
```
So, I applied shift amount first, then narrowed to 16-bit. As I used 'sxt' to widen the input, I kept using 'vasrwhsat' to match. The last commit should reflect this change. If I missed something here again, please let me know!